File: windows/JVoice.Core/Text/TextProcessor.cs
using System.Text;
using System.Text.RegularExpressions;
using JVoice.Core.Models;

namespace JVoice.Core.Text;

/// Post-processing pipeline: tone styles, filler removal, exact custom-word
/// corrections, hallucination-sentinel stripping. Faithful port of TextProcessor.swift.
public static class TextProcessor
{
    public static readonly IReadOnlyDictionary<string, string> CorrectionDictionary = new Dictionary<string, string>
    {
        ["app kit"] = "AppKit",
        ["appkit"] = "AppKit",
        ["j voice"] = "JVoice",
        ["jvoice"] = "JVoice",
        ["keyboard shortcuts"] = "KeyboardShortcuts",
        ["keyboardshortcuts"] = "KeyboardShortcuts",
        ["mac os"] = "macOS",
        ["whisper kit"] = "WhisperKit",
        ["whisperkit"] = "WhisperKit",
    };

    public static string Process(
        string text,
        ToneStyle mode,
        IReadOnlyDictionary<string, string>? extraDictionary = null,
        bool removeFillerWords = false,
        IReadOnlyList<string>? vocabulary = null)
    {
        extraDictionary ??= new Dictionary<string, string>();
        vocabulary ??= Array.Empty<string>();

        string normalized = NormalizeWhitespace(text);
        string clean = removeFillerWords ? RemoveDisfluencies(normalized) : normalized;
        // Very Casual lowercases first so corrections (applied after) win over the lowering.
        string cased = mode == ToneStyle.VeryCasual ? clean.ToLowerInvariant() : clean;
        string corrected = ApplyCorrections(cased, extraDictionary);
        string phonetic = PhoneticMatcher.Correct(corrected, vocabulary);
        return Format(phonetic, mode);
    }

    public static string ApplyCorrections(string text, IReadOnlyDictionary<string, string>? extraDictionary = null)
    {
        extraDictionary ??= new Dictionary<string, string>();
        var combined = new Dictionary<string, string>(extraDictionary);
        foreach (var kv in CorrectionDictionary) combined[kv.Key] = kv.Value; // builtin wins on conflict

        string result = text;
        foreach (var kv in combined.OrderByDescending(kv => kv.Key.Length))
            result = ReplaceOccurrences(kv.Key, result, kv.Value);
        return result;
    }

    public static IReadOnlyDictionary<string, string> BuildUserDictionary(IReadOnlyList<string> words)
    {
        var dict = new Dictionary<string, string>();
        foreach (var word in words)
            foreach (var variant in SpokenVariants(word))
                if (!CorrectionDictionary.ContainsKey(variant))
                    dict[variant] = word;
        return dict;
    }

    public static IReadOnlyList<string> ExtractCorrections(string original, string corrected)
    {
        var originalWords = original.Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries);
        var correctedWords = corrected.Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries);

        var results = new List<string>();
        if (originalWords.Length == correctedWords.Length)
        {
            for (int i = 0; i < originalWords.Length; i++)
            {
                if (originalWords[i] == correctedWords[i]) continue;
                var stripped = TrimPunctuation(correctedWords[i]);
                if (stripped.Length > 0) results.Add(stripped);
            }
        }
        else
        {
            var originalExact = new HashSet<string>(originalWords);
            foreach (var word in correctedWords)
            {
                if (originalExact.Contains(word)) continue;
                var stripped = TrimPunctuation(word);
                if (stripped.Length > 0) results.Add(stripped);
            }
        }
        return results.Distinct().Where(s => s.Length > 1).ToList();
    }

    public static IReadOnlyList<string> SpokenVariants(string word)
    {
        var variants = new HashSet<string>();
        string lower = word.ToLowerInvariant();

        variants.Add(lower);
        variants.Add(lower.Replace(" ", ""));
        variants.Add(lower.Replace(".", "").Replace(" ", ""));
        variants.Add(lower.Replace(".", " "));

        var camel = new StringBuilder();
        for (int i = 0; i < word.Length; i++)
        {
            if (char.IsUpper(word[i]) && i > 0) camel.Append(' ');
            camel.Append(char.ToLowerInvariant(word[i]));
        }
        variants.Add(camel.ToString());
        variants.Add(camel.ToString().Replace(" ", ""));

        return variants.Where(v => v.Length > 0 && v != word).ToList();
    }

    public static string Format(string text, ToneStyle mode)
    {
        string trimmed = text.Trim();
        if (trimmed.Length == 0) return "";

        return mode switch
        {
            ToneStyle.Casual => RemoveTerminalPunctuation(trimmed),
            ToneStyle.Formal => EnsureTerminalPeriod(CapitalizeFirstCharacter(trimmed)),
            ToneStyle.VeryCasual => EnsureTerminalDotOrQuestion(CollapseRepeatedCommas(trimmed)),
            _ => trimmed,
        };
    }

    private static string NormalizeWhitespace(string text)
        => string.Join(" ", text.Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries));

    public static string RemoveDisfluencies(string text)
    {
        string stripped = Regex.Replace(text, @"\b(um+h?|uh+|er+|a+h+|hmm+)\b[,.]?\s*", "", RegexOptions.IgnoreCase);
        string result = NormalizeWhitespace(stripped.Trim());
        if (result.EndsWith(",")) result = result[..^1];
        return result;
    }

    /// Removes "[BLANK_AUDIO]"-style all-caps/underscore bracket sentinels.
    public static string StripDecoderArtifacts(string text)
    {
        string stripped = Regex.Replace(text, @"\[[A-Z_][A-Z_ ]*\]", " ");
        return NormalizeWhitespace(stripped).Trim();
    }

    /// Returns "" when the whole input is hallucination noise (silence sentinels).
    public static string RemoveWhisperHallucinations(string text)
    {
        string trimmed = text.Trim();
        if (trimmed.Length == 0) return "";
        if (trimmed.All(c => ".,;:!? ".Contains(c))) return "";
        string[] blanklike =
        {
            "[BLANK_TEXT]", "BLANK_TEXT",
            "Thanks for watching!", "Thanks for watching.",
            "Thank you.", "Thank you for watching.",
            "Subscribe to my channel", "Subscribe to my channel.",
            "Please subscribe to my channel.",
            "Bye.", "Bye!",
        };
        foreach (var p in blanklike)
            if (string.Equals(trimmed, p, StringComparison.OrdinalIgnoreCase))
                return "";
        return text;
    }

    private static string ReplaceOccurrences(string needle, string text, string replacement)
    {
        string pattern = PhrasePattern(needle);
        // MatchEvaluator => literal replacement (no .NET $-group substitution surprises).
        return Regex.Replace(text, pattern, _ => replacement, RegexOptions.IgnoreCase);
    }

    private static string PhrasePattern(string phrase)
    {
        var components = phrase.Split(' ', StringSplitOptions.RemoveEmptyEntries)
            .Select(Regex.Escape);
        return @"\b" + string.Join(@"\s+", components) + @"\b";
    }

    private static string RemoveTerminalPunctuation(string text)
    {
        int end = text.Length;
        while (end > 0 && ".!?".Contains(text[end - 1])) end--;
        return text[..end];
    }

    private static string CapitalizeFirstCharacter(string text)
    {
        if (text.Length == 0) return text;
        return char.ToUpperInvariant(text[0]) + text[1..];
    }

    private static string EnsureTerminalPeriod(string text)
    {
        if (text.Length == 0) return text;
        char last = text[^1];
        return ".!?".Contains(last) ? text : text + ".";
    }

    private static string CollapseRepeatedCommas(string text)
        => Regex.Replace(text, @"\s*,(?:\s*,)*\s*", ", ");

    private static string EnsureTerminalDotOrQuestion(string text)
    {
        int end = text.Length;
        while (end > 0 && (text[end - 1] == ' ' || text[end - 1] == ',')) end--;
        string result = text[..end];
        if (result.Length == 0) return result;
        char last = result[^1];
        if (last == '?' || last == '.') return result;
        if (last == '!') return result[..^1] + ".";
        return result + ".";
    }

    private static string TrimPunctuation(string s)
    {
        int start = 0, end = s.Length;
        while (start < end && char.IsPunctuation(s[start])) start++;
        while (end > start && char.IsPunctuation(s[end - 1])) end--;
        return s[start..end];
    }
}

File: Sources/JVoice/Services/Transcription/TextProcessor.swift
import Foundation

public struct TextProcessor: Sendable {
    public static let shared = TextProcessor()

    public static let correctionDictionary: [String: String] = [
        "app kit": "AppKit",
        "appkit": "AppKit",
        "j voice": "JVoice",
        "jvoice": "JVoice",
        "keyboard shortcuts": "KeyboardShortcuts",
        "keyboardshortcuts": "KeyboardShortcuts",
        "mac os": "macOS",
        "whisper kit": "WhisperKit",
        "whisperkit": "WhisperKit"
    ]

    public init() {}

    public func process(_ text: String, mode: AppMode) -> String {
        Self.process(text, mode: mode)
    }

    public static func process(_ text: String, mode: AppMode, extraDictionary: [String: String] = [:], removeFillerWords: Bool = false, vocabulary: [String] = []) -> String {
        let normalized = normalizeWhitespace(text)
        let clean = removeFillerWords ? removeDisfluencies(normalized) : normalized
        // Very Casual lowercases the sentence — but corrections must win over
        // the lowering, so lower *first* and correct after. Custom words keep
        // their exact casing in every tone.
        let cased = mode == .veryCasual ? clean.lowercased() : clean
        let corrected = applyCorrections(cased, extraDictionary: extraDictionary)
        let phonetic = PhoneticMatcher.correct(corrected, vocabulary: vocabulary)
        return format(phonetic, mode: mode)
    }

    public func transform(_ text: String, mode: AppMode) -> String {
        process(text, mode: mode)
    }

    public static func transform(_ text: String, mode: AppMode) -> String {
        process(text, mode: mode)
    }

    public static func applyCorrections(_ text: String, extraDictionary: [String: String] = [:]) -> String {
        var result = text
        let combined = extraDictionary.merging(correctionDictionary) { _, builtin in builtin }
        for (needle, replacement) in combined.sorted(by: { $0.key.count > $1.key.count }) {
            result = replaceOccurrences(of: needle, in: result, with: replacement)
        }
        return result
    }

    public static func buildUserDictionary(from words: [String]) -> [String: String] {
        var dict: [String: String] = [:]
        for word in words {
            for variant in spokenVariants(for: word) where correctionDictionary[variant] == nil {
                dict[variant] = word
            }
        }
        return dict
    }

    public static func extractCorrections(from original: String, corrected: String) -> [String] {
        let originalWords = original.components(separatedBy: .whitespaces).filter { !$0.isEmpty }
        let correctedWords = corrected.components(separatedBy: .whitespaces).filter { !$0.isEmpty }

        var results: [String] = []

        if originalWords.count == correctedWords.count {
            for (orig, corr) in zip(originalWords, correctedWords) where orig != corr {
                let stripped = corr.trimmingCharacters(in: .punctuationCharacters)
                if !stripped.isEmpty { results.append(stripped) }
            }
        } else {
            let originalExact = Set(originalWords)
            for word in correctedWords where !originalExact.contains(word) {
                let stripped = word.trimmingCharacters(in: .punctuationCharacters)
                if !stripped.isEmpty { results.append(stripped) }
            }
        }

        return Array(Set(results)).filter { $0.count > 1 }
    }

    public static func spokenVariants(for word: String) -> [String] {
        var variants: Set<String> = []
        let lower = word.lowercased()

        variants.insert(lower)
        variants.insert(lower.replacingOccurrences(of: " ", with: ""))
        variants.insert(lower.replacingOccurrences(of: ".", with: "").replacingOccurrences(of: " ", with: ""))
        variants.insert(lower.replacingOccurrences(of: ".", with: " "))

        var camelSplit = ""
        for (i, char) in word.enumerated() {
            if char.isUppercase && i > 0 {
                camelSplit += " "
            }
            camelSplit += String(char).lowercased()
        }
        variants.insert(camelSplit)
        variants.insert(camelSplit.replacingOccurrences(of: " ", with: ""))

        // Drop any variant that is itself a substring of the canonical word
        // (case-insensitive). A punctuated custom word like ".NET" otherwise
        // registers the bare "net" variant, whose \b…\b pattern then re-matches
        // the letter-run inside the already-inserted ".NET" replacement,
        // corrupting it into "..NET"/"...NET" (TRX-01). A variant that is a
        // substring of the word can never correct anything the word itself
        // wouldn't already, so removing it is safe and stops the self-overlap.
        return Array(variants).filter { !$0.isEmpty && $0 != word && !lower.contains($0) }
    }

    public static func format(_ text: String, mode: AppMode) -> String {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return "" }

        switch mode {
        case .casual:
            return removeTerminalPunctuation(trimmed)
        case .formal:
            let capitalized = capitalizeFirstCharacter(trimmed)
            return ensureTerminalPeriod(capitalized)
        case .veryCasual:
            // Lowercasing happens in `process` *before* the correction passes
            // (so corrections survive); re-lowering here would destroy them.
            let tidied = collapseRepeatedCommas(trimmed)
            return ensureTerminalDotOrQuestion(tidied)
        }
    }

    private static func normalizeWhitespace(_ text: String) -> String {
        text
            .components(separatedBy: .whitespacesAndNewlines)
            .filter { !$0.isEmpty }
            .joined(separator: " ")
    }

    public static func removeDisfluencies(_ text: String) -> String {
        let pattern = #"(?i)\b(um+h?|uh+|er+|a+h+|hmm+)\b[,.]?\s*"#
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return text }
        let range = NSRange(text.startIndex..<text.endIndex, in: text)
        let stripped = regex.stringByReplacingMatches(in: text, options: [], range: range, withTemplate: "")
        var result = normalizeWhitespace(stripped.trimmingCharacters(in: .whitespacesAndNewlines))
        if result.hasSuffix(",") { result.removeLast() }
        return result
    }

    /// Known non-speech sentinel keywords WhisperKit/Whisper emit inside
    /// brackets. A bracketed token only counts as an artifact if its content
    /// contains an underscore (e.g. "BLANK_AUDIO", "NOISE_1") OR matches one of
    /// these words. This keeps legitimate dictated tokens like "[A]", "[I]",
    /// "[II]", "[X]" (option labels, citation markers, Roman numerals) intact.
    private static let decoderSentinelKeywords: Set<String> = [
        "BLANK_AUDIO", "BLANK_TEXT", "MUSIC", "APPLAUSE", "NOISE", "SILENCE",
        "INAUDIBLE", "LAUGHTER", "SOUND", "CROSSTALK"
    ]

    /// Removes WhisperKit/Whisper special-token renderings that leak into the
    /// transcript on silence / non-speech — "[BLANK_AUDIO]", "[BLANK_TEXT]",
    /// "[MUSIC]", "[APPLAUSE]", etc. These are never user speech and can appear
    /// mid-transcript on pause-heavy dictation. Only bracketed tokens that are
    /// real decoder sentinels (underscore-bearing or in `decoderSentinelKeywords`)
    /// are stripped, so legitimate dictated labels like "[A]"/"[II]"/"[X]" survive.
    public static func stripDecoderArtifacts(_ text: String) -> String {
        guard let regex = try? NSRegularExpression(pattern: #"\[[A-Z_][A-Z0-9_ ]*\]"#) else { return text }
        let nsText = text as NSString
        let range = NSRange(location: 0, length: nsText.length)
        var result = text
        // Replace strippable matches back-to-front so earlier ranges stay valid.
        for match in regex.matches(in: text, options: [], range: range).reversed() {
            let bracketed = nsText.substring(with: match.range)
            let inner = bracketed.dropFirst().dropLast() // strip the surrounding [ ]
            let isSentinel = inner.contains("_") || decoderSentinelKeywords.contains(String(inner))
            guard isSentinel, let r = Range(match.range, in: result) else { continue }
            result.replaceSubrange(r, with: " ")
        }
        return normalizeWhitespace(result).trimmingCharacters(in: .whitespacesAndNewlines)
    }

    /// Strips known Whisper hallucination outputs that occur on near-silent or zero-content
    /// audio. Returns an empty string if the entire input is hallucination noise.
    public static func removeWhisperHallucinations(_ text: String) -> String {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        // Lone punctuation
        if trimmed.isEmpty { return "" }
        if trimmed.allSatisfy({ ".,;:!? ".contains($0) }) { return "" }
        // Whisper-specific sentinels (stored without terminal punctuation).
        let blanklikePatterns = [
            "[BLANK_TEXT]",
            "BLANK_TEXT",
            "Thanks for watching",
            "Thank you",
            "Thank you for watching",
            "Subscribe to my channel",
            "Please subscribe to my channel",
            "Bye",
        ]
        // The tone formatter rewrites terminal "."/"!"/"?" before this filter
        // runs (Casual strips it, Formal/Very-Casual add one), so compare with
        // any trailing terminal punctuation removed — otherwise a whole-transcript
        // hallucination leaks whenever the formatter dropped its punctuation.
        let core = removeTerminalPunctuation(trimmed)
        for pattern in blanklikePatterns {
            if core.caseInsensitiveCompare(pattern) == .orderedSame {
                return ""
            }
        }
        return text
    }

    private static func replaceOccurrences(of needle: String, in text: String, with replacement: String) -> String {
        let pattern = phrasePattern(for: needle)
        guard let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive]) else {
            return text
        }
        let range = NSRange(text.startIndex..<text.endIndex, in: text)
        let safeTemplate = NSRegularExpression.escapedTemplate(for: replacement)
        return regex.stringByReplacingMatches(in: text, options: [], range: range, withTemplate: safeTemplate)
    }

    private static func phrasePattern(for phrase: String) -> String {
        let components = phrase
            .split(separator: " ")
            .map { NSRegularExpression.escapedPattern(for: String($0)) }
        return #"\b"# + components.joined(separator: #"\s+"#) + #"\b"#
    }

    private static func removeTerminalPunctuation(_ text: String) -> String {
        var result = text
        while let last = result.last, ".!?".contains(last) {
            result.removeLast()
        }
        return result
    }

    private static func capitalizeFirstCharacter(_ text: String) -> String {
        guard let first = text.first else { return text }
        return String(first).uppercased() + text.dropFirst()
    }

    private static func ensureTerminalPeriod(_ text: String) -> String {
        guard let last = text.last else { return text }
        if ".!?".contains(last) { return text }
        return text + "."
    }

    /// Collapses runs of commas — and the whitespace around them — into a
    /// single ", " so a very-casual transcript separates clauses without
    /// piling up commas.
    private static func collapseRepeatedCommas(_ text: String) -> String {
        guard let regex = try? NSRegularExpression(pattern: #"\s*,(?:\s*,)*\s*"#) else { return text }
        let range = NSRange(text.startIndex..<text.endIndex, in: text)
        return regex.stringByReplacingMatches(in: text, options: [], range: range, withTemplate: ", ")
    }

    /// Guarantees the very-casual output ends on a dot or a question mark.
    /// An existing `?`/`.` is kept, a trailing `!` becomes `.`, and any
    /// dangling comma or whitespace separator is dropped before the period.
    private static func ensureTerminalDotOrQuestion(_ text: String) -> String {
        var result = text
        while let last = result.last, last == " " || last == "," {
            result.removeLast()
        }
        guard let last = result.last else { return result }
        if last == "?" || last == "." { return result }
        if last == "!" {
            result.removeLast()
            return result + "."
        }
        return result + "."
    }
}
